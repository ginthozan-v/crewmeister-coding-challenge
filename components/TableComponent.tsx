import { useMemo } from 'react';
import { useTable, usePagination } from 'react-table';
import style from './TableComponent.module.css';

const TableComponent = ({ tableColumns, tableData, isLoading, error }: any) => {
  const columns = useMemo(() => tableColumns, [tableColumns]);
  const data = useMemo(() => tableData, [tableData]);

  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    page,
    nextPage,
    previousPage,
    canPreviousPage,
    canNextPage,
    prepareRow,
    state: { pageIndex },
  } = useTable(
    {
      columns,
      data,
    },
    usePagination
  );
  return (
    <>
      <table className={style.table} {...getTableProps()}>
        <thead>
          {headerGroups.map((headerGroup, i) => (
            <tr {...headerGroup.getHeaderGroupProps()} key={i}>
              {headerGroup.headers.map((column, i) => (
                <th {...column.getHeaderProps()} key={i}>
                  {column.render('Header')}
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {isLoading && (
            <tr>
              <td colSpan={tableColumns.length} style={{ textAlign: 'center' }}>
                Loading...
              </td>
            </tr>
          )}

          {!isLoading &&
            page.length > 0 &&
            page.map((row: any, i: number) => {
              prepareRow(row);
              return (
                <tr {...row.getRowProps()} key={i}>
                  {row.cells.map((cell: any, i: number) => {
                    return (
                      <td {...cell.getCellProps()} key={i}>
                        {cell.render('Cell')}
                      </td>
                    );
                  })}
                </tr>
              );
            })}

          {!isLoading && !page.length && (
            <tr>
              <td colSpan={tableColumns.length} style={{ textAlign: 'center' }}>
                {error ? error : 'No results!'}
              </td>
            </tr>
          )}
        </tbody>
      </table>
      <div className={style.btnGroup}>
        <button onClick={() => previousPage()} disabled={!canPreviousPage}>
          {'<'}
        </button>
        <span>{pageIndex + 1}</span>
        <button onClick={() => nextPage()} disabled={!canNextPage}>
          {'>'}
        </button>
      </div>
    </>
  );
};

export default TableComponent;
