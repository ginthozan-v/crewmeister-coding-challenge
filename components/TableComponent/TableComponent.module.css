.table {
  background-color: white;
  border-collapse: collapse;
  width: 100%;
  table-layout: fixed;
}

.table td,
.table th {
  border: 1px solid #ddd;
  padding: 10px;
}

.table th {
  padding-top: 12px;
  padding-bottom: 12px;
  background-color: rgb(59, 59, 59);
  color: white;
  font-size: 15px;
}

.table td {
  font-size: 14px;
}

.table tr:hover {
  background-color: rgb(230, 230, 230);
}

.btnGroup {
  display: flex;
  align-items: center;
  justify-content: flex-end;
  gap: 2px;
  margin-top: 10px;
  width: 100%;
}

.btnGroup button {
  width: 30px;
  height: 30px;
  border: 0;
  background-color: rgb(231, 231, 231);
  cursor: pointer;
}

.btnGroup button:hover:enabled {
  background-color: rgb(209, 209, 209);
}

.btnGroup span {
  width: 30px;
  height: 30px;
  border: 0;
  background-color: rgb(231, 231, 231);
  display: flex;
  align-items: center;
  justify-content: center;
}
