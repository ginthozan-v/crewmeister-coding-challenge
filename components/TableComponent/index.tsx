import { useMemo } from 'react';
import { useTable, usePagination } from 'react-table';
import style from './TableComponent.module.css';

interface IDataObject {
  id: string;
  name: string;
  type: string;
  period: string;
  memberNote: string;
  status: string;
  admitterNote: string;
  startDate: string;
  endDate: string;
}

interface ITableObject {
  tableColumns: any;
  tableData: IDataObject[];
  isLoading: boolean;
  error: string;
}

const TableComponent = ({
  tableColumns,
  tableData,
  isLoading,
  error,
}: ITableObject) => {
  const columns = useMemo(() => tableColumns, [tableColumns]);
  const data = useMemo(() => tableData, [tableData]);

  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    page,
    nextPage,
    previousPage,
    canPreviousPage,
    canNextPage,
    prepareRow,
    state: { pageIndex },
  } = useTable(
    {
      columns,
      data,
    },
    usePagination
  );
  return (
    <div className={style.tableWrapper}>
      <table className={style.table} {...getTableProps()}>
        <thead>
          {headerGroups.map((headerGroup, i) => (
            <tr {...headerGroup.getHeaderGroupProps()} key={i}>
              {headerGroup.headers.map((column, i) => (
                <th {...column.getHeaderProps()} key={i}>
                  {column.render('Header')}
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {isLoading && (
            <tr>
              <td colSpan={tableColumns.length} style={{ textAlign: 'center' }}>
                Loading...
              </td>
            </tr>
          )}

          {!isLoading &&
            page.length > 0 &&
            page.map((row: any, i: number) => {
              prepareRow(row);
              return (
                <tr {...row.getRowProps()} key={i}>
                  {row.cells.map((cell: any, i: number) => {
                    return (
                      <td {...cell.getCellProps()} key={i}>
                        {cell.render('Cell')}
                      </td>
                    );
                  })}
                </tr>
              );
            })}

          {!isLoading && !page.length && (
            <tr>
              <td colSpan={tableColumns.length} style={{ textAlign: 'center' }}>
                {error ? error : 'No results!'}
              </td>
            </tr>
          )}
        </tbody>
      </table>
      <div className={style.btnGroup}>
        <button onClick={() => previousPage()} disabled={!canPreviousPage}>
          {'<'}
        </button>
        <span>{pageIndex + 1}</span>
        <button onClick={() => nextPage()} disabled={!canNextPage}>
          {'>'}
        </button>
      </div>
    </div>
  );
};

export default TableComponent;
