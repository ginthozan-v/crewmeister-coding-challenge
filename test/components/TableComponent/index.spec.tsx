import React from 'react';
import { render } from '@testing-library/react';
import TableComponent from '../../../components/TableComponent';

describe('I want to see a list of absences', () => {
  test('should match', () => {
    const { container } = render(
      <TableComponent
        tableColumns={[
          {
            Header: 'Member name',
            accessor: 'name',
          },
          {
            Header: 'Type of absence',
            accessor: 'type',
          },
        ]}
        tableData={[
          {
            id: '1',
            name: 'Max',
            type: 'vacation',
            period: '2 days',
            memberNote: 'Skiurlaub',
            status: 'requested',
            admitterNote: '1',
            startDate: '2021-02-20',
            endDate: '2021-02-21',
          },
        ]}
        isLoading={false}
        error=""
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('should match the options', () => {
    const view = render(
      <TableComponent
        tableColumns={[
          {
            Header: 'Member name',
            accessor: 'name',
          },
          {
            Header: 'Type of absence',
            accessor: 'type',
          },
        ]}
        tableData={[
          {
            id: '1',
            name: 'Max',
            type: 'vacation',
            period: '2 days',
            memberNote: 'Skiurlaub',
            status: 'requested',
            admitterNote: '1',
            startDate: '2021-02-20',
            endDate: '2021-02-21',
          },
        ]}
        isLoading={false}
        error=""
      />
    );
    expect(view.container).toMatchSnapshot();
  });
});
