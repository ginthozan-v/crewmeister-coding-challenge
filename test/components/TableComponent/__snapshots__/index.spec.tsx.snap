// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I want to see a list of absences should match 1`] = `
<div>
  <table
    class="table"
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colspan="1"
          role="columnheader"
        >
          Member name
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Type of absence
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Max
        </td>
        <td
          role="cell"
        >
          vacation
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="btnGroup"
  >
    <button
      disabled=""
    >
      &lt;
    </button>
    <span>
      1
    </span>
    <button
      disabled=""
    >
      &gt;
    </button>
  </div>
</div>
`;

exports[`I want to see a list of absences should match the options 1`] = `
<div>
  <table
    class="table"
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colspan="1"
          role="columnheader"
        >
          Member name
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Type of absence
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Max
        </td>
        <td
          role="cell"
        >
          vacation
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="btnGroup"
  >
    <button
      disabled=""
    >
      &lt;
    </button>
    <span>
      1
    </span>
    <button
      disabled=""
    >
      &gt;
    </button>
  </div>
</div>
`;
